contrib/platform/src/com/sun/jna/platform/win32/W32Service.java

```diff
@@ -236,22 +236,23 @@ void stopService(long timeout) {
                 // the previouos implementation queried the service status and
                 // failed if the application did not correctly update its state
                 while(status.dwCurrentState != Winsvc.SERVICE_STOPPED) {
-                        try {
-				Thread.sleep( status.dwWaitHint );
+			long msRemainingBeforeTimeout = timeout - (System.currentTimeMillis() - startTime);
+
+			if (msRemainingBeforeTimeout < 0) {
+				throw new RuntimeException(String.format("Service stop exceeded timeout time of %d ms", timeout));
+			}
+
+			int dwWaitTime = Math.min(sanitizeWaitTime(status.dwWaitHint), (int)msRemainingBeforeTimeout);
```

Comment:
This needs to be declared as a long. The user could pass in a timeout value > `Integer.MAX_VALUE`, so msRemainingBeforeTimeout could also be > `Integer.MAX_VALUE`. After the cast, the value will wrap around and you will get negative wait time.

Reply:
Agreed.  Done and committed.